node/src/components/linear_chain_sync.rs

```diff
@@ -262,20 +331,60 @@ impl<I: Clone + PartialEq + 'static> LinearChainSync<I> {
             }
             State::SyncingTrustedHash {
                 ref mut linear_chain,
+                ref mut current_block,
                 ..
-            } => linear_chain.pop(),
+            } => match linear_chain.pop() {
+                None => None,
+                Some(block) => {
+                    // Update `current_block` so that we can verify whether result of execution
+                    // matches the expected value.
+                    current_block.replace(block.clone());
+                    Some(block)
+                }
+            },
             State::SyncingDescendants {
                 ref mut linear_chain,
+                ref mut current_block,
                 ..
-            } => linear_chain.pop_front(),
+            } => match linear_chain.pop_front() {
+                None => None,
```

Comment:
> Should we panic here? We only call `fetch_next_block_deploys` after downloading a block, I think?

I don't put `panic` in places where the cause doesn't break the system. Even if it shouldn't happen but happens, it should be harmless.


> Actually, should `linear_chain` even be an `Option<BlockHash>` in the `SyncingDescendants` variant, because it always contains at most one block?

Good idea. Let me have a look.

Reply:
Done.